README.md

```diff
@@ -61,8 +61,111 @@ This repository contains the following stand-alone applications, which demonstra
+
+Our Tiny Typir language has only a few possible AstElements, namely NumberLiteral, StringLiteral, BinaryExpression, Variable, and AssignmentStatement. They are implemented in a very simple way, see for example our BinaryExpression:
+
+```
+class BinaryExpression extends AstElement {
+    constructor(
+        public left: AstElement,
+        public operator: string,
+        public right: AstElement,
+    ) { super(); }
+}
+```
+
+Feel free to check out the others in the [test code](/packages/typir/test/api-example.test.ts), but a little spoiler: no surprises there.
+
+Let's head into setting up the Typir type system and creating the primitive types for our NumberLiteral and StringLiteral, which is a one line of code job each, as we use the Typir's predefined Primitives factory service:
+
+```
+const typir = createTypirServices();
+
+const numberType = typir.factory.Primitives.create({ primitiveName: 'number', inferenceRules: node => node instanceof NumberLiteral });
+
+const stringType = typir.factory.Primitives.create({ primitiveName: 'string', inferenceRules: node => node instanceof StringLiteral });
+```
+
+Note that the inference rules are included in this. For the operators this is a bit longer, as we have to take care of the left and right operand and the operator of the binary expression, so we
+extract it:
+
+```
+const inferenceRule: InferOperatorWithMultipleOperands<BinaryExpression> = {
+    filter: node => node instanceof BinaryExpression,
+    matching: (node, operatorName) => node.operator === operatorName,
+    operands: node => [node.left, node.right],
+};
+```
+
+We wish to have two operators, the `+` operator, which should be overloaded to accept either two numbers to add or two strings to concatenate. This can be expressed with an array of signatures with different types for the operands and the return type of the operator. Furthermore, there is going to be a `-` operator with only one signature, as we need no overloading. Both refer to the inferenceRule we defined above:
+
+```
+typir.factory.Operators.createBinary({ name: '+', signatures: [
+    { left: numberType, right: numberType, return: numberType },
+    { left: stringType, right: stringType, return: stringType },
+], inferenceRule });
+typir.factory.Operators.createBinary({ name: '-', signatures: [{ left: numberType, right: numberType, return: numberType }], inferenceRule });
+```
+
+As we'd like to be able to convert number to strings implicitly, we add the following line. Note that this will for example make it possible to concatenate numbers and strings with the `+` operator, though it as no signature for a number and a string parameter.
+
+```
+typir.Conversion.markAsConvertible(numberType, stringType,'IMPLICIT_EXPLICIT');
+```
+
+Furthermore we can specify how Typir should infer the variable type. We decided that the type of the variable should be the type if its initial value.
```

Comment:
Typir internally considers the inference rules for primitives and operators as well, when recursively inferring the given AstElement.

Reply:
Good, I added this information.